### PR TITLE
HIPP-1674 - Update API start controller

### PR DESCRIPTION
**Pull request by victorarbuesmallada**

### CODE REVIEW

File: app/controllers/myapis/update/UpdateApiStartController.scala [hunk 1]
```diff
@@ -0,0 +1,111 @@
...
+  def startProduceApi(id: String): Action[AnyContent] = (identify  andThen apiAuth(id)).async {
+    implicit request => {
+      val userAnswers = UserAnswers(
+        id = request.identifierRequest.user.userId,
+        lastUpdated = clock.instant()
+      )
+
+      (for {
+        deploymentDetails <- getDeploymentDetails(id)
+        updatedAnswers <- updateAnswer(userAnswers, UpdateApiApiPage, deploymentDetails)
```
Comment (victorarbuesmallada) [inline app/controllers/myapis/update/UpdateApiStartController.scala:63]: not sure what to do with this one, should I check the answers for `UpdateApiApiPage` before pulling the deployment details?

File: app/controllers/myapis/update/UpdateApiStartController.scala [hunk 2]
```diff
@@ -0,0 +1,111 @@
...
+    }
+  }
+
+  private def getDeploymentDetails(apiId: String)(implicit r: Request[?]) =
+    for {
+      apiDetail <- EitherT.fromOptionF(apiHubService.getApiDetail(apiId),
+        errorResultBuilder.apiNotFound(apiId)
+      )
+      deploymentDetails <- EitherT.fromOptionF(apiHubService.getDeploymentDetails(apiDetail.publisherReference),
+        errorResultBuilder.apiNotFoundInApim(apiDetail)
```
Comment (PaulCDurham) [inline app/controllers/myapis/update/UpdateApiStartController.scala:97]: We don't handle exceptions in controllers. Let the error handler do that to make sure things like consistent logging take place.

File: app/controllers/myapis/update/UpdateApiStartController.scala [hunk 3]
```diff
@@ -0,0 +1,111 @@
...
+      )
+    } yield (apiDetail, deploymentDetails)
+
+  private def updateAnswer[A](userAnswers: UserAnswers, page: QuestionPage[A], answer: A)
+                             (implicit r: Request[?], w: Writes[A]) =
+    Future.fromTry(userAnswers.set(page, answer))
+          .attemptT
+          .leftMap(e => errorResultBuilder.internalServerError(e))
```
Comment (PaulCDurham) [inline app/controllers/myapis/update/UpdateApiStartController.scala:105]: We don't handle exceptions in controllers. Let the error handler do that to make sure things like consistent logging take place.

File: app/models/myapis/ApiDomainSubdomain.scala [hunk 4]
```diff
@@ -14,12 +14,12 @@
  * limitations under the License.
  */
 
-package models.myapis.produce
+package models.myapis
 
 import play.api.libs.json.{Format, Json}
 
-case class ProduceApiDomainSubdomain(domain: String, subDomain: String)
+case class ApiDomainSubdomain(domain: String, subDomain: String)
```
Comment (PaulCDurham) [inline app/models/myapis/ApiDomainSubdomain.scala:21]: Why rename it? If you want to make it common then we need to define how we want to do that and implement that across all types.

File: test/controllers/myapis/update/UpdateApiStartControllerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,269 @@
...
+
+    "must return a 404 if there are no API details" in {
+      val fixture = buildFixture(maybeApiDetail = None)
+
+      running(fixture.application) {
+        val request = FakeRequest(controllers.myapis.update.routes.UpdateApiStartController.startProduceApi("id"))
+        val result = route(fixture.application, request).value
+
+        status(result) mustBe NOT_FOUND
+      }
```
Comment (PaulCDurham) [inline test/controllers/myapis/update/UpdateApiStartControllerSpec.scala:175]: Could you add assertions on the actual message and check the page is valid HTML?

File: test/controllers/myapis/update/UpdateApiStartControllerSpec.scala [hunk 6]
```diff
@@ -0,0 +1,269 @@
...
+    "must return a 404 if there are no deployment details" in {
+      val fixture = buildFixture()
+      when(fixture.apiHubService.getDeploymentDetails(any)(any))
+        .thenReturn(Future.successful(None))
+
+      running(fixture.application) {
+        val request = FakeRequest(controllers.myapis.update.routes.UpdateApiStartController.startProduceApi("id"))
+        val result = route(fixture.application, request).value
+
+        status(result) mustBe NOT_FOUND
```
Comment (PaulCDurham) [inline test/controllers/myapis/update/UpdateApiStartControllerSpec.scala:187]: Could you add assertions on the actual message and check the page is valid HTML?

File: test/controllers/myapis/update/UpdateApiStartControllerSpec.scala [hunk 7]
```diff
@@ -0,0 +1,269 @@
...
+      }
+    }
+
+    "must return a 500 if there is an error while saving the answers" in {
```
Comment (PaulCDurham) [inline test/controllers/myapis/update/UpdateApiStartControllerSpec.scala:191]: We don't typically test 500s. That's the responsibility of the error handler.

**Review by PaulCDurham — CHANGES_REQUESTED** [review]

File: app/models/myapis/ApiDomainSubdomain.scala (hunk 4 above)
Comment (victorarbuesmallada) [reply app/models/myapis/ApiDomainSubdomain.scala:21]: I did that before the chat we had this morning, will revert it back

**Review by PaulCDurham — APPROVED** [review]

**Pull request merged**